File: edgequake/sdks/java/src/test/java/io/edgequake/sdk/FakeHttpClient.java
package io.edgequake.sdk;

import io.edgequake.sdk.internal.HttpHelper;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSession;
import java.lang.reflect.Field;
import java.net.Authenticator;
import java.net.CookieHandler;
import java.net.ProxySelector;
import java.net.http.HttpClient;
import java.net.http.HttpHeaders;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.concurrent.Flow;

/**
 * Fake HttpClient that intercepts HTTP calls and returns
 * pre-configured responses without network I/O.
 *
 * <p>Usage:</p>
 * <pre>{@code
 * var pair = FakeHttpClient.createTestHelper(config);
 * var http = pair[0];   // HttpHelper
 * var fake = pair[1];   // FakeHttpClient
 *
 * fake.respondWith("{\"status\":\"healthy\"}");
 * var result = new HealthService(http).check();
 * assertEquals("healthy", result.status);
 * }</pre>
 */
public class FakeHttpClient extends HttpClient {

    public record CapturedRequest(String method, String uri, String body) {}

    private final List<CapturedRequest> captured = new ArrayList<>();
    private String responseBody = "{}";
    private int responseCode = 200;
    private Exception shouldThrow = null;

    public void respondWith(String json) {
        respondWith(json, 200);
    }

    public void respondWith(String json, int code) {
        this.responseBody = json;
        this.responseCode = code;
        this.shouldThrow = null;
    }

    public void respondWithError(int code) {
        respondWithError(code, "{\"error\":\"mock error\"}");
    }

    public void respondWithError(int code, String body) {
        this.responseBody = body;
        this.responseCode = code;
        this.shouldThrow = null;
    }

    public void throwOnSend(Exception ex) {
        this.shouldThrow = ex;
    }

    public CapturedRequest lastRequest() {
        return captured.get(captured.size() - 1);
    }

    public List<CapturedRequest> allRequests() {
        return Collections.unmodifiableList(captured);
    }

    public void clear() {
        captured.clear();
    }

    // ── HttpClient overrides ─────────────────────────────────────────

    @Override
    @SuppressWarnings("unchecked")
    public <T> HttpResponse<T> send(HttpRequest request,
                                     HttpResponse.BodyHandler<T> handler) throws java.io.IOException {
        // Capture request body
        String bodyStr = null;
        if (request.bodyPublisher().isPresent()) {
            var pub = request.bodyPublisher().get();
            var sb = new StringBuilder();
            pub.subscribe(new Flow.Subscriber<ByteBuffer>() {
                @Override public void onSubscribe(Flow.Subscription s) { s.request(Long.MAX_VALUE); }
                @Override public void onNext(ByteBuffer item) {
                    var arr = new byte[item.remaining()];
                    item.get(arr);
                    sb.append(new String(arr));
                }
                @Override public void onError(Throwable t) {}
                @Override public void onComplete() {}
            });
            try { Thread.sleep(10); } catch (InterruptedException ignored) {}
            bodyStr = sb.length() > 0 ? sb.toString() : null;
        }

        captured.add(new CapturedRequest(request.method(), request.uri().toString(), bodyStr));

        if (shouldThrow != null) {
            if (shouldThrow instanceof java.io.IOException ioe) throw ioe;
            throw new java.io.IOException(shouldThrow);
        }

        final var code = responseCode;
        final var body = responseBody;

        return (HttpResponse<T>) new HttpResponse<String>() {
            @Override public int statusCode() { return code; }
            @Override public HttpHeaders headers() {
                return HttpHeaders.of(
                    Map.of("content-type", List.of("application/json")),
                    (k, v) -> true);
            }
            @Override public String body() { return body; }
            @Override public java.net.URI uri() { return request.uri(); }
            @Override public HttpRequest request() { return request; }
            @Override public Optional<HttpResponse<String>> previousResponse() { return Optional.empty(); }
            @Override public Optional<SSLSession> sslSession() { return Optional.empty(); }
            @Override public Version version() { return Version.HTTP_1_1; }
        };
    }

    @Override
    public <T> CompletableFuture<HttpResponse<T>> sendAsync(HttpRequest request,
                                                             HttpResponse.BodyHandler<T> handler) {
        try {
            return CompletableFuture.completedFuture(send(request, handler));
        } catch (Exception e) {
            return CompletableFuture.failedFuture(e);
        }
    }

    @Override
    public <T> CompletableFuture<HttpResponse<T>> sendAsync(HttpRequest request,
                                                             HttpResponse.BodyHandler<T> handler,
                                                             HttpResponse.PushPromiseHandler<T> pushHandler) {
        return sendAsync(request, handler);
    }

    @Override public Optional<CookieHandler> cookieHandler() { return Optional.empty(); }
    @Override public Optional<Duration> connectTimeout() { return Optional.of(Duration.ofSeconds(30)); }
    @Override public Redirect followRedirects() { return Redirect.NEVER; }
    @Override public Optional<ProxySelector> proxy() { return Optional.empty(); }
    @Override public SSLContext sslContext() { return null; }
    @Override public SSLParameters sslParameters() { return new SSLParameters(); }
    @Override public Optional<Authenticator> authenticator() { return Optional.empty(); }
    @Override public Version version() { return Version.HTTP_1_1; }
    @Override public Optional<Executor> executor() { return Optional.empty(); }

    // ── Factory ──────────────────────────────────────────────────────

    /**
     * Create an HttpHelper with a fake HTTP client for unit testing.
     * Uses reflection to inject the FakeHttpClient into HttpHelper's private httpClient field.
     *
     * @return Object[]{HttpHelper, FakeHttpClient}
     */
    public static Object[] createTestHelper(EdgeQuakeConfig config) {
        var helper = new HttpHelper(config);
        var fake = new FakeHttpClient();

        try {
            Field field = HttpHelper.class.getDeclaredField("httpClient");
            field.setAccessible(true);
            field.set(helper, fake);
        } catch (Exception e) {
            throw new RuntimeException("Failed to inject FakeHttpClient", e);
        }

        return new Object[]{helper, fake};
    }
}

File: edgequake/sdks/kotlin/src/test/kotlin/io/edgequake/sdk/FakeHttpClient.kt
package io.edgequake.sdk

import io.edgequake.sdk.internal.HttpHelper
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.net.http.HttpResponse.BodyHandler
import java.nio.ByteBuffer
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Flow
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLParameters
import java.net.Authenticator
import java.net.CookieHandler
import java.net.ProxySelector
import java.time.Duration
import java.util.Optional

/**
 * Fake HttpClient for unit testing.
 *
 * Captures requests and returns pre-configured responses
 * without network I/O.
 */
class FakeHttpClient : HttpClient() {

    data class CapturedRequest(
        val method: String,
        val uri: String,
        val body: String?
    )

    val captured = mutableListOf<CapturedRequest>()
    var responseBody: String = "{}"
    var responseCode: Int = 200
    var shouldThrow: Exception? = null

    fun respondWith(json: String, code: Int = 200) {
        responseBody = json
        responseCode = code
        shouldThrow = null
    }

    fun respondWithError(code: Int, body: String = """{"error":"mock error"}""") {
        responseBody = body
        responseCode = code
        shouldThrow = null
    }

    fun throwOnSend(ex: Exception) {
        shouldThrow = ex
    }

    fun lastRequest(): CapturedRequest = captured.last()
    fun clear() { captured.clear() }

    @Suppress("UNCHECKED_CAST")
    override fun <T> send(
        request: HttpRequest,
        responseBodyHandler: BodyHandler<T>
    ): HttpResponse<T> {
        // Capture request body
        val bodyStr = request.bodyPublisher().map { pub ->
            val sb = StringBuilder()
            pub.subscribe(object : Flow.Subscriber<ByteBuffer> {
                override fun onSubscribe(s: Flow.Subscription) { s.request(Long.MAX_VALUE) }
                override fun onNext(item: ByteBuffer) {
                    val arr = ByteArray(item.remaining())
                    item.get(arr)
                    sb.append(String(arr))
                }
                override fun onError(throwable: Throwable) {}
                override fun onComplete() {}
            })
            // Give subscriber a moment
            Thread.sleep(10)
            sb.toString().ifEmpty { null }
        }.orElse(null)

        captured.add(CapturedRequest(request.method(), request.uri().toString(), bodyStr))

        shouldThrow?.let { throw it }

        // Build a fake response
        return object : HttpResponse<T> {
            override fun statusCode() = responseCode
            override fun headers() = java.net.http.HttpHeaders.of(mapOf("content-type" to listOf("application/json"))) { _, _ -> true }
            override fun body() = responseBody as T
            override fun uri() = request.uri()
            override fun request() = request
            override fun previousResponse() = Optional.empty<HttpResponse<T>>()
            override fun sslSession() = Optional.empty<javax.net.ssl.SSLSession>()
            override fun version() = Version.HTTP_1_1
        }
    }

    override fun <T> sendAsync(
        request: HttpRequest,
        responseBodyHandler: BodyHandler<T>
    ) = CompletableFuture.completedFuture(send(request, responseBodyHandler))

    override fun <T : Any?> sendAsync(
        request: HttpRequest,
        responseBodyHandler: BodyHandler<T>,
        pushPromiseHandler: HttpResponse.PushPromiseHandler<T>
    ): CompletableFuture<HttpResponse<T>> = sendAsync(request, responseBodyHandler)

    // Required abstract method implementations
    override fun cookieHandler() = Optional.empty<CookieHandler>()
    override fun connectTimeout() = Optional.of(Duration.ofSeconds(30))
    override fun followRedirects() = Redirect.NEVER
    override fun proxy() = Optional.empty<ProxySelector>()
    override fun sslContext(): SSLContext = SSLContext.getDefault()
    override fun sslParameters() = SSLParameters()
    override fun authenticator() = Optional.empty<Authenticator>()
    override fun version() = Version.HTTP_1_1
    override fun executor() = Optional.empty<java.util.concurrent.Executor>()
    override fun newWebSocketBuilder() = throw UnsupportedOperationException()
}

/**
 * Create an HttpHelper with a fake HTTP client for unit testing.
 *
 * Uses reflection to inject the FakeHttpClient since httpClient is @PublishedApi internal.
 */
fun createTestHelper(config: EdgeQuakeConfig = EdgeQuakeConfig()): Pair<HttpHelper, FakeHttpClient> {
    val helper = HttpHelper(config)
    val fakeClient = FakeHttpClient()

    // Inject fake client via reflection
    val field = HttpHelper::class.java.getDeclaredField("httpClient")
    field.isAccessible = true
    field.set(helper, fakeClient)

    return Pair(helper, fakeClient)
}
